File: src/mcp/fs.ts
/**
 * R2 filesystem MCP (§7).
 *
 * Exposes the persistent file store through tools ONLY. Every call is
 * authorized server-side against the session manifest's folder grants with
 * separate read / write / delete verbs. No raw blob access is ever given to
 * an agent.
 */
import { z } from "zod";
import { createMcpServer, type McpServer, type McpRuntime } from "./context.js";
import { checkFsOp, normalizePath } from "../acl/grants.js";

export function createFsMcpServer(): McpServer {
  return createMcpServer("r2-fs", [
    {
      name: "fs.list",
      description: "List entries under a folder path.",
      inputSchema: z.object({ path: z.string().default("/") }),
      handler: async (rt, { path }) => {
        assert(rt, "list", path);
        const entries = await rt.services.files.list(rt.manifest.projectId, path);
        return {
          entries: entries.map((e) => ({
            path: e.path,
            type: e.type,
            size: e.size,
          })),
        };
      },
    },
    {
      name: "fs.read",
      description: "Read a file's UTF-8 content.",
      inputSchema: z.object({ path: z.string() }),
      handler: async (rt, { path }) => {
        assert(rt, "read", path);
        const { file, content } = await rt.services.files.read(rt.manifest.projectId, path);
        return {
          path: file.path,
          mime: file.mime,
          size: file.size,
          content: content.toString("utf8"),
        };
      },
    },
    {
      name: "fs.write",
      description: "Write (create or replace) a UTF-8 text file. Persists beyond the session.",
      inputSchema: z.object({ path: z.string(), content: z.string(), mime: z.string().optional() }),
      handler: async (rt, { path, content, mime }) => {
        assert(rt, "write", path);
        const file = await rt.services.files.write(rt.manifest.projectId, path, content, mime);
        return { ok: true, path: file.path, size: file.size };
      },
    },
    {
      name: "fs.mkdir",
      description: "Validate a folder path (folders are virtual; requires write grant).",
      inputSchema: z.object({ path: z.string() }),
      handler: async (rt, { path }) => {
        assert(rt, "mkdir", path);
        return { ok: true, path: normalizePath(path) };
      },
    },
    {
      name: "fs.delete",
      description: "Delete a file. Requires canDelete on the folder grant.",
      inputSchema: z.object({ path: z.string() }),
      handler: async (rt, { path }) => {
        assert(rt, "delete", path);
        await rt.services.files.delete(rt.manifest.projectId, path);
        return { ok: true };
      },
    },
  ]);
}

function assert(
  rt: McpRuntime,
  op: "list" | "read" | "write" | "mkdir" | "delete",
  path: string,
): void {
  const res = checkFsOp(rt.manifest.filesystemGrants, op, path);
  if (!res.ok) throw new Error(`fs.${op} denied: ${res.reason}`);
}

File: agentos/mcp/fs.py
"""R2 filesystem MCP. Port of src/mcp/fs.ts."""
from agentos.acl.grants import check_fs_op, normalize_path
from agentos.mcp.context import McpServer, McpTool, create_mcp_server


def create_fs_mcp_server() -> McpServer:
    def _assert(rt, op, path):
        ok, reason = check_fs_op(rt.manifest.filesystemGrants, op, path)
        if not ok:
            raise PermissionError(f"fs.{op} denied: {reason}")

    async def fs_list(rt, args):
        path = args.get("path", "/")
        _assert(rt, "list", path)
        entries = await rt.services.files.list(rt.manifest.projectId, path)
        return {"entries": [{"path": e["path"], "type": e["type"], "size": e["size"]} for e in entries]}

    async def fs_read(rt, args):
        _assert(rt, "read", args["path"])
        file, content = await rt.services.files.read(rt.manifest.projectId, args["path"])
        return {"path": file["path"], "mime": file["mime"], "size": file["size"],
                "content": content.decode(errors="replace")}

    async def fs_write(rt, args):
        _assert(rt, "write", args["path"])
        file = await rt.services.files.write(
            rt.manifest.projectId, args["path"], args["content"], args.get("mime")
        )
        return {"ok": True, "path": file["path"], "size": file["size"]}

    async def fs_mkdir(rt, args):
        _assert(rt, "mkdir", args["path"])
        return {"ok": True, "path": normalize_path(args["path"])}

    async def fs_delete(rt, args):
        _assert(rt, "delete", args["path"])
        await rt.services.files.delete(rt.manifest.projectId, args["path"])
        return {"ok": True}

    return create_mcp_server("r2-fs", [
        McpTool("fs.list", "List entries under a folder path.", fs_list),
        McpTool("fs.read", "Read a file's UTF-8 content.", fs_read),
        McpTool("fs.write", "Write (create or replace) a UTF-8 text file.", fs_write),
        McpTool("fs.mkdir", "Validate a folder path (requires write grant).", fs_mkdir),
        McpTool("fs.delete", "Delete a file (requires delete grant).", fs_delete),
    ])
